www/js/lib/uiUtil.js

```diff
@@ -225,6 +225,77 @@ define(rqDef, function() {
+            // Create a Message Channel
+            var channel = new MessageChannel();
+            // Handler for recieving message reply from service worker
+            channel.port1.onmessage = function (event) {
+                var cacheNames = event.data;
+                if (cacheNames.error) return;
+                else {
+                    caches.keys().then(function (keyList) {
+                        if (keyList.length < 3) {
+                            updateAlert.style.display = 'none';
+                            appstate.updateNeeded = false;
+                            return;
+                        }
+                        keyList.forEach(function (key) {
+                            if (key === cacheNames.app || key === cacheNames.assets) return;
+                            // If we get here, then there is a cache key that does not match our version, i.e. a PWA-in-waiting
+                            appstate.updateNeeded = true;
+                            updateAlert.style.display = 'block';
+                            document.getElementById('persistentMessage').innerHTML = 'Version ' + key.replace(/^[^\d]+/, '') +
+                                ' is ready to install. (Re-launch app to install.)';
+                        });
+                    });
+                }
+            };
+            if (navigator.serviceWorker.controller) navigator.serviceWorker.controller.postMessage({
+                action: 'getCacheNames'
+            }, [channel.port2]);
+        }
+    }
+    if (updateAlert) updateAlert.querySelector('button[data-hide]').addEventListener('click', function () {
+        updateAlert.style.display = 'none';
+    });
+
+    /**
+     * Checks if a server is accessible by attempting to load a test image from the server
+     * @param {String} imageSrc The full URI of the image
+     * @param {any} onSuccess A function to call if the image can be loaded
+     * @param {any} onError A function to call if the image cannot be loaded
+     */
+     function checkServerIsAccessible(imageSrc, onSuccess, onError) {
+        var image = new Image();
+        image.onload = onSuccess;
+        image.onerror = onError;
+        image.src = imageSrc;
+    }
+
+    /**
+     * Show or hide the spinner together with a message
+     * @param {Boolean} show True to show the spinner, false to hide it 
+     * @param {String} message A message to display, or hide the message if null 
+     */
+    function spinnerDisplay(show, message) {
+        var searchingArticles = document.getElementById('searchingArticles');
+        var spinnerMessage = document.getElementById('cachingAssets');
+        if (show) searchingArticles.style.display = 'block';
+        else searchingArticles.style.display = 'none';
+        if (message) {
+            spinnerMessage.innerHTML = message;
+            spinnerMessage.style.display = 'block';
+        } else {
+            spinnerMessage.innerHTML = 'Caching assets...';
```

Comment:
It's a bit strange to have this string here

Reply:
I'm not sure these changes belong in this PR, I can't remember why they've changed. I'll check. Maybe some incidental code cleanup.

Reply:
These functions do belong here. I have to show the spinner when the user requests to use Service Worker in the extension, while the app is checking that it has access to the server. This can take a few seconds and can be disconcerting otherwise. So I tidied up the spinnerDisplay routine in the app and made a function for it, which allows simplification of code at that point (and elsewhere in the app).